File: python/codeql/command_line_injection/benign_edge.py
import subprocess

from flask import Flask, request


app = Flask(__name__)
COMMANDS = {
    "version": ["python3", "--version"],
    "whoami": ["whoami"],
}


@app.get("/command-safe")
def command_safe():
    command = COMMANDS.get(request.args.get("command", "version"), COMMANDS["version"])
    return subprocess.check_output(command, text=True)

File: javascript/codeql/command_line_injection/benign_edge.js
const express = require("express");
const childProcess = require("child_process");

const app = express();
const commands = new Map([["version", ["node", ["--version"]]], ["whoami", ["whoami", []]]]);

app.get("/command-safe", (req, res) => {
  const [command, args] = commands.get(String(req.query.command || "version")) || commands.get("version");
  childProcess.execFile(command, args, (error, stdout) => {
    res.status(error ? 500 : 200).send(stdout);
  });
});
